### PR TITLE
feat(py): FirestoreSessionStore (google-cloud plugin)

### PR DESCRIPTION
**Pull request by huangjeff5**
Adds a Firestore-backed agent session store. PR 1 of 3 in the FastAPI-agents stack; PR2 (serve_agent/serve_flow) and PR3 (sample) build on this.

### CODE REVIEW

**Comment by huangjeff5** [comment]
@gemini-code-assist review

File: py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+            asyncio.run_coroutine_threadsafe(cleanup(), loop)
+
+            def unsubscribe_safely() -> None:
+                if watch_holder:
+                    watch_holder[0].unsubscribe()
+
+            loop.call_soon_threadsafe(unsubscribe_safely)
+
+        watch_holder.append(ref.on_snapshot(on_snapshot))
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py]: Updated _start_listener to check for .on_snapshot and fall back to initializing a synchronous firestore.Client when given an AsyncDocumentReference. Also corrected the callback signature to (doc_snapshots, changes, read_time).

File: py/plugins/google-cloud/src/genkit/plugins/google_cloud/session_store/firestore.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+            asyncio.run_coroutine_threadsafe(cleanup(), loop)
+
+            def unsubscribe_safely() -> None:
+                if watch_holder:
+                    watch_holder[0].unsubscribe()
+
+            loop.call_soon_threadsafe(unsubscribe_safely)
+
+        watch_holder.append(ref.on_snapshot(on_snapshot))
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/plugins/google-cloud/src/genkit/plugins/google_cloud/session_store/firestore.py]: Updated _start_listener to check for .on_snapshot and fall back to initializing a synchronous firestore.Client when given an AsyncDocumentReference. Also corrected the callback signature to (doc_snapshots, changes, read_time).

File: py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+    async def _maybe_update_pointer(self, session_id: str, snapshot_id: str, *, is_new: bool) -> None:
+        ref = self._pointer_ref(session_id)
+        existing = await ref.get()
+        pointer = existing.to_dict() if existing.exists else None
+        if is_new or not pointer or pointer.get('currentSnapshotId') == snapshot_id:
+            await ref.set({
+                'currentSnapshotId': snapshot_id,
+                'updatedAt': datetime.now(timezone.utc).isoformat(),
+            })
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py]: Wrapped read-then-write in an async transaction (@firestore.async_transactional) and switched timestamp to firestore.SERVER_TIMESTAMP.

File: py/plugins/google-cloud/src/genkit/plugins/google_cloud/session_store/firestore.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+    async def _maybe_update_pointer(self, session_id: str, snapshot_id: str, *, is_new: bool) -> None:
+        ref = self._pointer_ref(session_id)
+        existing = await ref.get()
+        pointer = existing.to_dict() if existing.exists else None
+        if is_new or not pointer or pointer.get('currentSnapshotId') == snapshot_id:
+            await ref.set({
+                'currentSnapshotId': snapshot_id,
+                'updatedAt': datetime.now(timezone.utc).isoformat(),
+            })
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/plugins/google-cloud/src/genkit/plugins/google_cloud/session_store/firestore.py]: Wrapped read-then-write in an async transaction (@firestore.async_transactional) and switched timestamp to firestore.SERVER_TIMESTAMP.

File: py/packages/genkit-google-cloud/tests/firestore_session_store_test.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    terminal_doc = MagicMock()
+    terminal_doc.exists = True
+    terminal_doc.to_dict.return_value = {
+        'snapshotId': 'snap-sub',
+        'sessionId': 'sess-1',
+        'createdAt': '2026-07-03T00:00:00Z',
+        'status': 'aborted',
+    }
+
+    captured_cb[0](terminal_doc)
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/tests/firestore_session_store_test.py]: Updated mock invocation to pass [terminal_doc], None, None matching the 3-arg callback signature.

File: py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+        snapshot_path_prefix: Callable[[], str] | None = None,
+        reject_ambiguous_session: bool = False,
+    ) -> None:
+        """Initialize the Firestore session store."""
+        self._client = client or firestore.AsyncClient()
+        self._collection = collection
+        self._prefix_fn = snapshot_path_prefix or (lambda: DEFAULT_PREFIX)
+        self._reject_ambiguous = reject_ambiguous_session
+        self._lock = asyncio.Lock()
+        self._subs: Subs = {}
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py]: Added self._sync_client placeholder in __init__ for lazy initialization and reuse across listeners.

File: py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+            sync_client = firestore.Client(
+                project=self._client.project,
+                credentials=getattr(self._client, '_credentials', None),
+                database=getattr(self._client, '_database', getattr(self._client, 'database', None)),
+            )
+            prefix = self._prefix_fn()
+            ref = (
+                sync_client.collection(self._collection).document(prefix).collection('snapshots').document(snapshot_id)
+            )
+        loop = asyncio.get_event_loop()
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py]: Updated _start_listener to lazily initialize and reuse self._sync_client, and replaced get_event_loop() with get_running_loop().

**Comment by huangjeff5** [comment]
@gemini-code-assist review

File: py/plugins/google-cloud/src/genkit/plugins/google_cloud/session_store/firestore.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+        snapshot_path_prefix: Callable[[], str] | None = None,
+        reject_ambiguous_session: bool = False,
+    ) -> None:
+        """Initialize the Firestore session store."""
+        self._client = client or firestore.AsyncClient()
+        self._collection = collection
+        self._prefix_fn = snapshot_path_prefix or (lambda: DEFAULT_PREFIX)
+        self._reject_ambiguous = reject_ambiguous_session
+        self._lock = asyncio.Lock()
+        self._subs: Subs = {}
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/plugins/google-cloud/src/genkit/plugins/google_cloud/session_store/firestore.py]: Removed obsolete plugins/google-cloud directory from legacy layout (migrated to packages/genkit-google-cloud in commit 31f2dde7a).

File: py/plugins/google-cloud/src/genkit/plugins/google_cloud/session_store/firestore.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+            sync_client = firestore.Client(
+                project=self._client.project,
+                credentials=getattr(self._client, '_credentials', None),
+                database=getattr(self._client, '_database', getattr(self._client, 'database', None)),
+            )
+            prefix = self._prefix_fn()
+            ref = (
+                sync_client.collection(self._collection).document(prefix).collection('snapshots').document(snapshot_id)
+            )
+        loop = asyncio.get_event_loop()
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/plugins/google-cloud/src/genkit/plugins/google_cloud/session_store/firestore.py]: Removed obsolete plugins/google-cloud directory from legacy layout (migrated to packages/genkit-google-cloud in commit 31f2dde7a).

File: py/plugins/google-cloud/src/genkit/plugins/google_cloud/session_store/firestore.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+            status_val = data.get('status')
+            if status_val is None:
+                return
+            try:
+                status = SnapshotStatus(status_val)
+            except ValueError:
+                return
+            loop.call_soon_threadsafe(lambda: _notify(self._subs, snapshot_id, status))
+
+        ref.on_snapshot(on_snapshot)
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/plugins/google-cloud/src/genkit/plugins/google_cloud/session_store/firestore.py]: Implemented terminal status unsubscription and cleanup.

File: py/plugins/google-cloud/src/genkit/plugins/google_cloud/session_store/firestore.py [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+            return next_snapshot
+
+    async def on_snapshot_status_change(self, snapshot_id: str) -> asyncio.Queue[SnapshotStatus | None]:
+        """Subscribe to status changes for a session snapshot."""
+        async with self._lock:
+            current = await asyncio.to_thread(self._read_snapshot_sync, snapshot_id)
+            q = await _subscribe(self._subs, snapshot_id, current)
+        # Firestore listener keeps cross-instance abort working for detached turns.
+        self._start_listener(snapshot_id, q)
+        return q
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/plugins/google-cloud/src/genkit/plugins/google_cloud/session_store/firestore.py]: Added check to ensure only a single Firestore listener starts per snapshot ID.

File: py/plugins/google-cloud/src/genkit/plugins/google_cloud/session_store/firestore.py [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+        # Prefix scan via collection listing — fine for moderate session counts.
+        snaps: list[SessionSnapshot] = []
+        for doc in self._snapshots_col().stream():
+            try:
+                snap = SessionSnapshot.model_validate(doc.to_dict())
+            except (ValueError, TypeError):
+                continue
+            if _session_id_of(snap) == session_id:
+                snaps.append(snap)
+        return snaps
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/plugins/google-cloud/src/genkit/plugins/google_cloud/session_store/firestore.py]: Replaced full collection streaming with direct query on sessionId field.

File: py/plugins/google-cloud/src/genkit/plugins/google_cloud/session_store/firestore.py [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+
+StateT = TypeVar('StateT')
+
+DEFAULT_COLLECTION = 'genkit-sessions'
+DEFAULT_PREFIX = 'global'
+
+
+def _sanitize(value: Any) -> Any:  # noqa: ANN401
+    """Drop None members so Firestore accepts the payload."""
+    return json.loads(json.dumps(value, default=str))
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/plugins/google-cloud/src/genkit/plugins/google_cloud/session_store/firestore.py]: Updated payload sanitization to recursively strip None values.

File: py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+
+            loop.call_soon_threadsafe(lambda: _notify(self._subs, snapshot_id, None))
+
+            async def cleanup() -> None:
+                async with self._lock:
+                    self._subs.pop(snapshot_id, None)
+
+            loop.create_task(cleanup())
+            if watch_holder:
+                watch_holder[0].unsubscribe()
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py]: Updated listener unsubscription and cleanup scheduling to be thread-safe via loop.call_soon_threadsafe.

File: py/plugins/google-cloud/src/genkit/plugins/google_cloud/session_store/firestore.py [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+
+            loop.call_soon_threadsafe(lambda: _notify(self._subs, snapshot_id, None))
+
+            async def cleanup() -> None:
+                async with self._lock:
+                    self._subs.pop(snapshot_id, None)
+
+            loop.create_task(cleanup())
+            if watch_holder:
+                watch_holder[0].unsubscribe()
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/plugins/google-cloud/src/genkit/plugins/google_cloud/session_store/firestore.py]: Updated listener unsubscription and cleanup scheduling to be thread-safe.

**Comment by huangjeff5** [comment]
@gemini-code-assist review

**Comment by huangjeff5** [comment]
@gemini-code-assist review

**Comment by huangjeff5** [comment]
@gemini-code-assist review

File: py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+        reject_ambiguous_session: bool = False,
+    ) -> None:
+        """Initialize the Firestore session store."""
+        self.client = client or firestore.AsyncClient()
+        self.collection = collection
+        self.prefix_fn = snapshot_path_prefix or (lambda: DEFAULT_PREFIX)
+        self.reject_ambiguous = reject_ambiguous_session
+        self.lock = asyncio.Lock()
+        self.subs: Subs = {}
+        self.sync_client: firestore.Client | None = None
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py]: Updated lock initialization to be lazy via a lock property.

File: py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py [hunk 17]
```diff
@@ -0,0 +1,261 @@
...
+
+    async def on_snapshot_status_change(self, snapshot_id: str) -> asyncio.Queue[SnapshotStatus | None]:
+        """Subscribe to status changes for a session snapshot."""
+        async with self.lock:
+            current = await self.read_snapshot(snapshot_id)
+            is_first = snapshot_id not in self.subs
+            q = await subscribe(self.subs, snapshot_id, current)
+            # Firestore listener keeps cross-instance abort working for detached turns.
+            if is_first:
+                self.start_listener(snapshot_id)
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py]: Wrapped start_listener call in try/except to clean up subscription on failure.

File: py/packages/genkit-google-cloud/tests/firestore_session_store_test.py [hunk 18]
```diff
@@ -0,0 +1,122 @@
+# Copyright 2026 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+"""Tests for FirestoreSessionStore."""
+
+import asyncio
+from typing import Any
+from unittest.mock import AsyncMock, MagicMock
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/tests/firestore_session_store_test.py]: Imported patch from unittest.mock.

File: py/packages/genkit-google-cloud/tests/firestore_session_store_test.py [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+        'status': 'aborted',
+    }
+
+    captured_cb[0]([terminal_doc], None, None)
+    await asyncio.sleep(0.05)
+
+    assert await queue.get() == SnapshotStatus.ABORTED
+    assert await queue.get() is None
+    watch_mock.unsubscribe.assert_called_once()
+    assert 'snap-sub' not in store.subs
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/tests/firestore_session_store_test.py]: Updated test to delete on_snapshot from mock_doc_ref and patch firestore.Client to test fallback path.

**Comment by huangjeff5** [comment]
@gemini-code-assist review

File: py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py [hunk 20]
```diff
@@ -0,0 +1,272 @@
...
+                    message="FirestoreSessionStore requires 'sessionId' on the snapshot.",
+                )
+
+            await self.write_snapshot(next_snapshot)
+            await self.maybe_update_pointer(session_id, sid, is_new=existing is None)
+            notify(self.subs, sid, next_snapshot.status)
+            return next_snapshot
+
+    async def on_snapshot_status_change(self, snapshot_id: str) -> asyncio.Queue[SnapshotStatus | None]:
+        """Subscribe to status changes for a session snapshot."""
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py:162]: Optimized listener startup to skip snapshots in terminal status.

File: py/packages/genkit-google-cloud/tests/firestore_session_store_test.py [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+    with patch('genkit_google_cloud.session_store.firestore.firestore.Client', return_value=mock_sync_client):
+        queue = await store.on_snapshot_status_change('snap-sub')
+        assert await queue.get() == SnapshotStatus.PENDING
+        assert len(captured_cb) == 1
+
+        terminal_doc = MagicMock()
+        terminal_doc.exists = True
+        terminal_doc.to_dict.return_value = {
+            'snapshotId': 'snap-sub',
+            'sessionId': 'sess-1',
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/tests/firestore_session_store_test.py:122]: Added test case verifying CAS pointer updates when appending to existing sessions.

File: py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+            sid = next_snapshot.snapshot_id
+            session_id = session_id_of(next_snapshot)
+            if not session_id:
+                raise GenkitError(
+                    status='INVALID_ARGUMENT',
+                    message="FirestoreSessionStore requires 'sessionId' on the snapshot.",
+                )
+
+            await self.write_snapshot(next_snapshot)
+            await self.maybe_update_pointer(session_id, sid, is_new=existing is None)
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py]: Updated save_snapshot to pass parent_snapshot_id to maybe_update_pointer for CAS CAS concurrency.

File: py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+            if is_new or not pointer or pointer.get('currentSnapshotId') == snapshot_id:
+                transaction.set(
+                    ref,
+                    {
+                        'currentSnapshotId': snapshot_id,
+                        'updatedAt': firestore.SERVER_TIMESTAMP,
+                    },
+                )
+
+        await update_in_transaction(transaction)
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py]: Replaced is_new with parent_snapshot_id in maybe_update_pointer to implement optimistic CAS pointer updates.

File: py/packages/genkit-google-cloud/tests/firestore_session_store_test.py [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+            session_id='sess-456',
+            created_at='2026-07-03T00:00:01Z',
+        )
+
+    saved = await store.save_snapshot(None, save_fn)
+    assert saved is not None
+    assert saved.session_id == 'sess-456'
+    assert isinstance(saved.snapshot_id, str)
+    mock_doc_ref.set.assert_called()
+    mock_transaction.set.assert_called_once()
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/tests/firestore_session_store_test.py:69]: Added CAS append verification to test_firestore_session_store_save_and_get.

**Comment by huangjeff5** [comment]
@gemini-code-assist review

File: py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py [hunk 25]
```diff
@@ -0,0 +1,272 @@
...
+        """Save or update a session snapshot in Firestore."""
+        async with self.lock:
+            existing = await self.read_snapshot(snapshot_id) if snapshot_id is not None else None
+            next_snapshot = apply_save(existing, snapshot_id, fn)
+            if next_snapshot is None:
+                return None
+
+            sid = next_snapshot.snapshot_id
+            session_id = session_id_of(next_snapshot)
+            if not session_id:
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py:149]: Implemented compare-and-swap pointer updates using parent_snapshot_id in commit 853650fce.

File: py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py [hunk 26]
```diff
@@ -0,0 +1,272 @@
...
+        payload = snapshot.model_dump(by_alias=True, exclude_none=True, mode='json')
+        await self.snapshot_ref(snapshot.snapshot_id).set(payload)
+
+    async def read_pointer(self, session_id: str) -> str | None:
+        """Read the current snapshot ID pointer for a session from Firestore."""
+        doc = await self.pointer_ref(session_id).get()
+        if not doc.exists:
+            return None
+        current = (doc.to_dict() or {}).get('currentSnapshotId')
+        return current if isinstance(current, str) else None
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py:198]: Implemented compare-and-swap pointer updates using parent_snapshot_id in commit 853650fce.

File: py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py [hunk 27]
```diff
@@ -0,0 +1,279 @@
...
+    def start_listener(self, snapshot_id: str) -> None:
+        """Start a Firestore real-time listener for status changes on a snapshot."""
+        ref = self.snapshot_ref(snapshot_id)
+        if not hasattr(ref, 'on_snapshot'):
+            if self.sync_client is None:
+                self.sync_client = firestore.Client(
+                    project=self.client.project,
+                    credentials=getattr(self.client, '_credentials', None),
+                    database=getattr(self.client, '_database', getattr(self.client, 'database', None)),
+                )
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py:246]: Passed client_options from AsyncClient to fallback sync Client.

File: py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py [hunk 28]
```diff
@@ -0,0 +1,279 @@
...
+
+            async def cleanup() -> None:
+                async with self.lock:
+                    self.subs.pop(snapshot_id, None)
+
+            asyncio.run_coroutine_threadsafe(cleanup(), loop)
+
+            def unsubscribe_safely() -> None:
+                if watch_holder:
+                    watch_holder[0].unsubscribe()
```
[the remark this answers is not in the document]
Comment (huangjeff5) [reply py/packages/genkit-google-cloud/src/genkit_google_cloud/session_store/firestore.py]: Ran watch unsubscribe in executor to avoid blocking asyncio event loop.

**Comment by huangjeff5** [comment]
@gemini-code-assist review

**Comment by huangjeff5** [comment]
# `FirestoreSessionStore` End-to-End & Bug-Bash Audit Report

**Date**: 2026-07-09  
**Target Package**: `genkit-google-cloud` (`FirestoreSessionStore`)  
**Target Repository & PR**: `genkit-ai/genkit` PR #5685 (commit `a6697d5a9`)  
**Test Environments**:
1. In-memory deterministic simulation (`MockFirestoreEngine` supporting compare-and-swap transactions, queries, and watch streams).
2. Live Google Cloud Firestore service (project: `aim-testing`, collections: `genkit-e2e-sessions` / `genkit-e2e-sessions-pointers`).

---

## 1. Executive Summary

We conducted an exhaustive bug-bash and end-to-end (E2E) integration audit of `FirestoreSessionStore`, evaluating its conformance with the Genkit Python Agents runtime (`Agent`, `Session`, `load_session`, `AgentChat`, tool calls, detached background tasks, interrupts, and real-time status subscriptions).

During edge-case stress testing, we discovered and remediated **2 critical bugs** in `FirestoreSessionStore`. Both test suites (`firestore_session_store_bug_bash_test.py` and `firestore_session_store_e2e_test.py`) have been added to the permanent test suite and pushed to PR #5685. All 83 tests across `genkit-google-cloud` pass cleanly in 13.48s alongside 100% compliance with `ruff`, `pyright`, `pyrefly`, and `ty`.

---

## 2. Bugs Discovered & Remediated

### Bug 1: `AsyncDocumentReference.on_snapshot` `NotImplementedError` Crash
- **Symptom**: Calling `on_snapshot_status_change(snapshot_id)` when the store was initialized with an `AsyncClient` crashed with `NotImplementedError` from `google.cloud.firestore_v1.base_document.py:359`.
- **Root Cause**: In the Google Cloud Firestore Python SDK, `AsyncDocumentReference` defines `on_snapshot(self, callback)` solely to raise `NotImplementedError`, as real-time watch listeners are currently only implemented on synchronous clients. Our fallback condition checked `if not hasattr(ref, 'on_snapshot'):`, which evaluated to `True` on `AsyncDocumentReference` and attempted to execute the unimplemented method directly.
- **Remediation**: Updated `start_listener` in `firestore.py` to check `if isinstance(self.client, firestore.AsyncClient) or not hasattr(ref, 'on_snapshot'):`. When an async client is provided, the store automatically initializes and caches `self.sync_client` with matching credentials and client options to manage real-time watch streams safely.

### Bug 2: Terminal Snapshot Listener Queue Hang
- **Symptom**: When `on_snapshot_status_change` subscribed to a snapshot that was already in a terminal status (`COMPLETED`, `ABORTED`, `FAILED`, `EXPIRED`), consumer loops (such as `AgentRuntime.watch_snapshot_abort`) hung indefinitely awaiting queue items.
- **Root Cause**: In `on_snapshot_status_change`, if a snapshot was already terminal, `start_listener(snapshot_id)` was intentionally bypassed to conserve gRPC resources. However, `subscribe` only seeded the queue with the initial status (e.g., `ABORTED`). Because the listener was never started, `notify(self.subs, snapshot_id, None)` was never invoked to send the terminating sentinel (`None`). Consumers expecting `None` to signal stream termination hung forever on `await queue.get()`.
- **Remediation**: Updated `on_snapshot_status_change` to inspect the initial status upon subscription. If `current.status in TERMINAL_STATUSES`, the store immediately puts `None` into the queue following the status seed, pops `snapshot_id` from `self.subs`, and returns without opening a gRPC watch stream.

---

## 3. In-Memory Bug-Bash Suite (`firestore_session_store_bug_bash_test.py`)

We built `MockFirestoreEngine`, an in-memory simulation of Firestore supporting transactional Compare-And-Swap (CAS) operations, field queries (`where().stream()`), and real-time watcher callbacks. We verified 7 distinct edge-case scenarios:

| Test Case | Scenario Exercised | Verification & Assertions |
| :--- | :--- | :--- |
| `test_bug_bash_multiturn_linear_chat_pointer_advancement` | Multiturn linear chat (Turns 1, 2, 3). | Verified `save_snapshot` appends link `parent_id` sequentially. Verified pointer document `currentSnapshotId` advances atomicaly from Turn 1 -> Turn 2 -> Turn 3. Verified `get_snapshot(session_id=...)` returns Turn 3 tip. |
| `test_bug_bash_branching_and_forked_history` | History forking (Branch A tip vs Branch B fork from Root). | Verified appending Branch B to Root leaves pointer intact at Branch A tip (CAS isolation). Verified un-pointed fallback with `reject_ambiguous_session=False` returns latest leaf chronologically, while `reject_ambiguous_session=True` raises `GenkitError(FAILED_PRECONDITION, 'branching snapshots')`. |
| `test_bug_bash_cas_concurrent_append_race_condition` | Concurrent append race condition (Turn 2X and Turn 2Y both targeting Turn 1). | Simulated simultaneous transaction execution. Verified Turn 2X commits and updates pointer to `turn-2x`. Verified Turn 2Y CAS check (`pointer == 'turn-1'`) fails and leaves pointer at `turn-2x` without overwriting. |
| `test_bug_bash_inplace_update_vs_append` | In-place snapshot update vs appending new turns. | Verified calling `save_snapshot(turn1_id, fn)` updates status from `PENDING` to `COMPLETED` idempotently without altering snapshot ID or creating duplicate records. Verified pointer remains stable. |
| `test_bug_bash_realtime_status_listener_and_cleanup` | Real-time status subscription and thread-safe listener cleanup. | Verified queue receives live transitions (`PENDING` -> `COMPLETED` -> `None`). Verified `watch.unsubscribe()` is executed in an executor and subscription is popped from `store.subs`. |
| `test_bug_bash_terminal_status_optimization` | Subscribing to an already-terminal snapshot (`ABORTED`). | Verified `start_listener` is bypassed. Verified queue receives `ABORTED` followed immediately by `None` sentinel. |
| `test_bug_bash_error_handling_and_invalid_selectors` | Invalid selector combinations and missing metadata. | Verified saving without `sessionId` raises `GenkitError(INVALID_ARGUMENT)`. Verified calling `get_snapshot` with 0 or 2 selectors raises `GenkitError(INVALID_ARGUMENT, 'exactly one of')`. |

---

## 4. Live Google Cloud Firestore E2E Suite (`firestore_session_store_e2e_test.py`)

We executed end-to-end integration tests against the live Google Cloud Firestore service on project `aim-testing` using tenant-isolated prefixes (`test-e2e-{uuid4().hex[:8]}`). All test data was verified directly via Firestore DocumentReference queries and wiped clean during teardown.

### Workflow 1: E2E Agent Workflow & Tool Execution (`test_firestore_session_store_e2e_agent_workflow_and_tools`)
- **Execution**: Defined `e2eAgent` with a custom `calculator` tool (`input_val * 2`). Executed Turn 1 (`"Calculate 21 * 2"`) and Turn 2 (`"What was the answer again?"`) via `AgentChat`.
- **Database Verification**:
  - Queried `genkit-e2e-sessions/global/snapshots/{turn1_id}` directly in Firestore. Verified `status: 'completed'`, `sessionId`, and exact message history structure: 4 parts durably stored (`role: 'user'`, `role: 'model'` with `ToolRequestPart`, `role: 'tool'` with `ToolResponsePart` output `42`, and `role: 'model'` final reply).
  - Queried `genkit-e2e-sessions-pointers/global/pointers/{session_id}` in Firestore. Verified `currentSnapshotId` pointed to Turn 1 after Turn 1, and atomically advanced to Turn 2 after Turn 2.
  - Verified Turn 2 document in Firestore correctly referenced `parentId: {turn1_id}` and contained all 6 accumulated message parts.

### Workflow 2: E2E Detach, Abort, and Real-Time gRPC Watch Stream (`test_firestore_session_store_e2e_detach_abort_and_watch`)
- **Execution**: Configured `slowAgent` with a 4-second delay. Initiated a detached background turn via `task = await chat.detach('Run slow background analysis')`. Subscribed to real-time status updates via `e2e_store.on_snapshot_status_change(task.snapshot_id)` over a live gRPC watch stream against Google Cloud Firestore.
- **Database & Stream Verification**:
  - Verified initial status received from watch stream and direct DB query was `PENDING`.
  - Executed client-side abort via `await task.abort()`.
  - Verified live gRPC watch stream pushed `SnapshotStatus.ABORTED` followed by sentinel `None`.
  - Queried Firestore document directly; verified durable status transitioned to `aborted`.
  - Verified awaiting the aborted turn raised `asyncio.CancelledError`.
  - Sent follow-up prompt on the same chat; verified conversation resumed cleanly from the last completed turn without stranding history.

### Workflow 3: E2E Branching, CAS Isolation, & Strict Ambiguous Rejection (`test_firestore_session_store_e2e_branching_and_cas_isolation`)
- **Execution**: Executed Root turn -> Branch A turn. Verified pointer in Firestore pointed to Branch A tip. Called `agent.load_chat(snapshot_id=root_id)` and executed Branch B turn (`"Branch B prompt"`).
- **Database Verification**:
  - Queried Branch B document in Firestore; verified `parentId` matched Root ID.
  - Verified pointer in Firestore remained safely at Branch A tip due to CAS isolation (`currentSnapshotId == root_id` evaluated to False when pointer was at Branch A).
  - Deleted pointer document in Firestore to simulate a legacy/un-pointed session with multiple branch tips.
  - Initialized `strict_store` with `reject_ambiguous_session=True`. Attempted to load session via `strict_agent.load_chat(session_id=session_id)`.
  - Verified Firestore query detected 2 leaf documents and raised `GenkitError(FAILED_PRECONDITION)` matching `'branching snapshots'`.

---

## 5. Verification & Test Run Logs

```bash
uv run --no-sync pytest packages/genkit-google-cloud/tests -v --no-cov
```

```
============================= test session starts ==============================
platform darwin -- Python 3.12.9, pytest-9.0.2, pluggy-1.6.0
rootdir: /Users/huangjeff/Desktop/genkit-firestore-session/py
configfile: pyproject.toml
plugins: anyio-4.12.1, mock-3.15.1, Faker-40.4.0, asyncio-1.3.0, typeguard-4.5.0, cov-7.0.0
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=session, asyncio_default_test_loop_scope=function
collecting ... collected 83 items

packages/genkit-google-cloud/tests/firestore_session_store_bug_bash_test.py::test_bug_bash_multiturn_linear_chat_pointer_advancement PASSED [  1%]
packages/genkit-google-cloud/tests/firestore_session_store_bug_bash_test.py::test_bug_bash_branching_and_forked_history PASSED [  2%]
packages/genkit-google-cloud/tests/firestore_session_store_bug_bash_test.py::test_bug_bash_cas_concurrent_append_race_condition PASSED [  3%]
packages/genkit-google-cloud/tests/firestore_session_store_bug_bash_test.py::test_bug_bash_inplace_update_vs_append PASSED [  4%]
packages/genkit-google-cloud/tests/firestore_session_store_bug_bash_test.py::test_bug_bash_realtime_status_listener_and_cleanup PASSED [  6%]
packages/genkit-google-cloud/tests/firestore_session_store_bug_bash_test.py::test_bug_bash_terminal_status_optimization PASSED [  7%]
packages/genkit-google-cloud/tests/firestore_session_store_bug_bash_test.py::test_bug_bash_error_handling_and_invalid_selectors PASSED [  8%]
packages/genkit-google-cloud/tests/firestore_session_store_e2e_test.py::test_firestore_session_store_e2e_agent_workflow_and_tools PASSED [  9%]
packages/genkit-google-cloud/tests/firestore_session_store_e2e_test.py::test_firestore_session_store_e2e_detach_abort_and_watch PASSED [ 10%]
packages/genkit-google-cloud/tests/firestore_session_store_e2e_test.py::test_firestore_session_store_e2e_branching_and_cas_isolation PASSED [ 12%]
... [73 existing unit tests passed] ...
======================= 83 passed, 6 warnings in 13.48s ========================
```

All formatting (`ruff format --check`), linter rules (`ruff check`), and type checkers (`pyright`, `pyrefly`, `ty`) pass with 0 errors across `genkit-google-cloud`.
